app/helpers/foreman_host_reports/reports_dashboard_helper.rb

```diff
@@ -0,0 +1,74 @@
+module ForemanHostReports
+  module ReportsDashboardHelper
+    def host_reports_searchable_links(name, search, counter, format)
+      content_tag :li, :style => "margin-bottom: 5px" do
+        content_tag(:span, sanitize('&nbsp;'), :class => 'label', :style => "background-color: #{host_reports_report_color[counter]}") +
+          sanitize('&nbsp;') +
+          link_to(name, hosts_path(:search => search), :class => "dashboard-links") +
+          content_tag(:span, send(counter, format), class: 'pull-right')
+      end
+    end
+
+    def host_reports_report_color
+      {
+        :change_hosts => "#4572A7",
+        :failure_hosts => "#AA4643",
+        :nochange_hosts => "#DB843D",
+        :disabled_hosts => "#92A8CD",
+      }
+    end
+
+    def host_reports_get_overview(options = {})
+      format = options[:format]
+      state_labels = {
+        change_hosts: _('Change'),
+        failure_hosts: _('Failure'),
+        nochange_hosts: _('No Change'),
+        disabled_hosts: _('Disabled alerts'),
+      }
+
+      {
+        data: state_labels.map { |key, label| [label, send(key, format), host_reports_report_color[key]] },
+        searchUrl: hosts_path(search: '~VAL~'),
+        searchFilters: state_labels.each_with_object({}) do |(key, filter), filters|
+          filters[filter] = send(key, format)
+        end,
+      }
+    end
+
+    def latest_reports
+      hosts = Host.authorized(:view_hosts, Host)
+      HostReport.authorized(:view_host_reports).my_reports
+                .where(host_id: hosts.pluck(:id))
+                .reorder(:last_report => :desc)
+                .limit(9)
+                .preload(:host)
+    end
+
+    def latest_reports?
+      latest_reports.limit(1).present?
+    end
+
+    def host_reports_translated_header(shortname, longname)
+      "<th style='width:85px; class='ca'><span class='small' title='' data-original-title='#{longname}'>#{shortname}</span></th>"
+    end
+
+    private
+
+    def disabled_hosts(_format = nil)
+      Host.authorized(:view_hosts, Host).where(:enabled => false).count
+    end
+
+    def change_hosts(format)
+      Host.joins(:host_reports).where("format = ? and change > 0", HostReport.formats[format]).uniq.count
+    end
+
+    def nochange_hosts(format)
+      Host.joins(:host_reports).where("format = ? and nochange > 0", HostReport.formats[format]).uniq.count
+    end
+
+    def failure_hosts(format)
+      Host.joins(:host_reports).where("format = ? and failure > 0", HostReport.formats[format]).uniq.count
+    end
```

Comment:
So I spent quite some time trying to figure out a single huge query that would do all three selects via SQL list with subselect but then I ditched the idea completely it got extremely complicated once taxonomy got involved. This is a tested way how to fetch the data:

```ruby
HostReport.authorized(:view_host_reports).my_reports.where("format = ? and nochange > 0", HostReport.formats[:puppet]).reorder('').group(:host_id).maximum(:id).count
```

I tested this with taxonomy in mind too, after I created UserA that only belongs to OrgA which only has a one host:

```
User.current = User.find(6)

[34] pry(main)> HostReport.authorized(:view_host_reports).my_reports.where("format = ? and nochange > 0", HostReport.formats[:puppet]).reorder('').group(:host_id).maximum(:id).count

2022-02-21T18:08:44 [D|per|] checking permission view_host_reports for class HostReport
2022-02-21T18:08:44 [D|per|] organization_ids: []
2022-02-21T18:08:44 [D|per|] location_ids: []
2022-02-21T18:08:44 [D|per|] filter with role_id: 12 limited: false search:  taxonomy_search:
2022-02-21T18:08:44 [D|per|] checking permission view_hosts for class Host
2022-02-21T18:08:44 [D|per|] organization_ids: []
2022-02-21T18:08:44 [D|per|] location_ids: []
2022-02-21T18:08:44 [D|per|] filter with role_id: 2 limited: false search:  taxonomy_search:
 | filter with role_id: 3 limited: false search:  taxonomy_search:
 | filter with role_id: 6 limited: false search:  taxonomy_search:
 | filter with role_id: 7 limited: false search:  taxonomy_search:
 | filter with role_id: 8 limited: false search:  taxonomy_search:
 | filter with role_id: 9 limited: false search:  taxonomy_search:
 | filter with role_id: 15 limited: false search:  taxonomy_search:
 | filter with role_id: 16 limited: false search:  taxonomy_search:
 | filter with role_id: 24 limited: false search:  taxonomy_search:
 | filter with role_id: 25 limited: false search:  taxonomy_search:
 | filter with role_id: 27 limited: false search:  taxonomy_search:
=> 1
```

It correctly shows just one host while there are multiple in other orgs in my case. This is the change we need to do.

Reply:
I am not sure I fully understand what you did. 
The problem you mentioned is:

> But there is one problem - we need to count only the last reports, the query appears to count across all reports per host. What we want is to count hosts for which their last reports have failures/unchanged/changed.

Does this fix it? because I don't see something related to host.last_report in the query.

Reply:
> Does this fix it?

Yes it does, note the group and max aggregate functions.